server/src/main/java/org/opensearch/rest/pagination/IndexBasedPaginationStrategy.java

```diff
@@ -0,0 +1,191 @@
+                        requestedPageEndIndexNumber,
+                        clusterState.metadata().indices().get(sortedIndicesList.get(requestedPageEndIndexNumber - 1)).getCreationDate(),
+                        queryStartTime,
+                        sortedIndicesList.get(requestedPageEndIndexNumber - 1)
+                    ).generateEncryptedToken(),
+                paginatedElement
+            );
+        }
+    }
+
+    @Override
+    @Nullable
+    public PaginatedQueryResponse getPaginatedQueryResponse() {
+        return paginatedQueryResponse;
+    }
+
+    @Override
+    public List<String> getElementsFromRequestedToken() {
+        return Objects.isNull(indicesFromRequestedToken) ? new ArrayList<>() : indicesFromRequestedToken;
+    }
+
+    private int getRequestedPageIndexStartNumber(
+        final IndexStrategyTokenParser requestedTokenParser,
+        final List<String> sortedIndicesList,
+        final ClusterState clusterState,
+        final String sortOrder
+    ) {
+        if (Objects.isNull(requestedTokenParser)) {
+            return 0;
+        }
+        int requestedPageStartIndexNumber = Math.min(requestedTokenParser.posToStartPage, sortedIndicesList.size());
+        if (requestedPageStartIndexNumber > 0
+            && !Objects.equals(sortedIndicesList.get(requestedPageStartIndexNumber - 1), requestedTokenParser.nameOfLastRespondedIndex)) {
+            // case denoting an already responded index has been deleted while the paginated queries are being executed
+            // find the index whose creation time is just after/before (based on sortOrder) the index which was last responded.
+            while (requestedPageStartIndexNumber > 0) {
+                long creationTimeOfIndex = clusterState.metadata()
+                    .indices()
+                    .get(sortedIndicesList.get(requestedPageStartIndexNumber - 1))
+                    .getCreationDate();
+                if ((!DESCENDING_SORT_PARAM_VALUE.equals(sortOrder)
+                    && creationTimeOfIndex < requestedTokenParser.creationTimeOfLastRespondedIndex)
+                    || (DESCENDING_SORT_PARAM_VALUE.equals(sortOrder)
+                        && creationTimeOfIndex > requestedTokenParser.creationTimeOfLastRespondedIndex)) {
+                    break;
+                }
+                requestedPageStartIndexNumber--;
+            }
+        }
+        return requestedPageStartIndexNumber;
+    }
+
+    /**
+     * TokenParser to be used by {@link IndexBasedPaginationStrategy}.
+     * Token would like: IndexNumberToStartTheNextPageFrom + $ + CreationTimeOfLastRespondedIndex + $ +
+     * QueryStartTime + $ + NameOfLastRespondedIndex
+     */
+    public static class IndexStrategyTokenParser {
+
+        private final int posToStartPage;
+        private final long creationTimeOfLastRespondedIndex;
+        private final long queryStartTime;
+        private final String nameOfLastRespondedIndex;
```

Comment:
Can we add documentation on the significance of each of the fields and how it is used to retrieve the page ?